### PR TITLE
Refactored active month roster filtering, made month and withdrawn filters mutually exclusive

### PR DESCRIPTION
**Pull request by vonbearshark**
## Background
This is trickier than it lets on. There is ambiguity in the difference between `active`, `withdrawn`, and `enrolled` with month filtering. The Issue has captured most of the context, but please help me poke the cases a little more

## GitHub Issue
#1187 

## Associated PRs
N / A

## Validation Plan
Created a ~user~ child on the border, as outlined in the ticket, ensured this change now includes that new ~user~child in the roster.

## Automated Testing
- [ ] All unit tests are passing.
- [ ] All e2e tests are passing.

## Additional Context
I saw snapshot changes in other PRs. Am I missing a snapshot generation step?

### CODE REVIEW

File: client/src/containers/Roster/Roster.tsx [hunk 1]
```diff
@@ -80,31 +80,32 @@ const Roster: React.FC = () => {
...
-  const rosterIsEmpty = (allChildren || []).length === 0;
-  const displayChildren = query.withdrawn ? withdrawnChildren : activeChildren;
-  const isSingleSiteView = query.site ? true : false;
+  } = filterChildrenByWithdrawn(allChildren);
+
+  // Withdrawn and Month filters are mutually exclusive, active is the default in the absence of either
+  const displayChildren = query.withdrawn
+    ? withdrawnChildren
+    : queryMonth
+    ? allChildren
```
Comment (jlhogan) [inline client/src/containers/Roster/Roster.tsx:98]: names of these variables could be more informative -- to my brain it's not obvious that "allChildren" is the month filtered children. I think this could be simplified by the comment i left in the GH issue, where the "active"(default) roster is just the roster with query month = this month

File: src/controllers/children/get.ts [hunk 2]
```diff
@@ -71,16 +70,19 @@ export const getChildren = async (
...
-
-      // filter out children with no qualifying enrollments
-      return c.enrollments && c.enrollments.length;
-    });
+    const monthEnd = activeMonth.endOf('month');
+    // Return only children with active enrollment during that month
+    return children
+      .map((child) => ({
+        ...child,
+        // Return enrollments that started before the filter and either did not end, or ended after the end of the filter month
```
Comment (jlhogan) [inline src/controllers/children/get.ts:78]: I'm not grokking why adding this month end check ensures children who were previously missing are now included halp

**Comment by jlhogan** [comment]
re snapshots -- if you run the FE tests with `-u` flag, any snapshots that need to be updated will. Because the tests have already gone green here, and you didnt add any snapshot tests, you don't need to update/create any snapshots!

File: src/controllers/children/get.ts [hunk 3]
```diff
@@ -71,16 +70,19 @@ export const getChildren = async (
...
+        enrollments: child.enrollments?.filter(
+          (e) =>
+            e.entry?.isSameOrBefore(monthEnd) &&
+            (!e.exit || e.exit.isAfter(monthEnd))
+        ),
+      }))
+      .filter((c) => c.enrollments?.length); // filter out children with no qualifying enrollments
```
[the remark this answers is not in the document]
Comment (jlhogan) [reply src/controllers/children/get.ts:85]: im 99.999% sure prettier will enforce syntax as it is

File: src/controllers/children/get.ts (hunk 2 above)
Comment (vonbearshark) [reply src/controllers/children/get.ts:78]: Ugh I had a comment queued from last night that didn't post https://github.com/ctoec/data-collection/pull/1231#discussion_r589897482

File: src/controllers/children/get.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (vonbearshark) [reply src/controllers/children/get.ts:85]: Unfortunately, yes. Also I need to configure my local dev setup or something because I would prefer the prettier on save, while this kind of snuck up on me at commit time, I think? Maybe I am imagining things.

[the remark this answers is not in the document]
Comment (vonbearshark) [reply src/controllers/children/get.ts:85]: I _can_ break it like

```
const childrenDuringMonth = children.map((child) => ({
  // ...
 }));
 return childrenDuringMonth.filter((c) => c.enrollments?.length);
 ```
 
 no strong preferences there

File: src/controllers/children/get.ts [hunk 4]
```diff
@@ -71,16 +71,19 @@ export const getChildren = async (
...
+    const monthEnd = activeMonth.endOf('month');
+    // Return only children with active enrollment during that month
+    return children
+      .map((child) => ({
+        ...child,
+        // Return enrollments that started before the filter and either did not end, or ended after the end of the filter month
+        enrollments: child.enrollments?.filter(
+          (e) =>
+            e.entry?.isSameOrBefore(monthEnd) &&
+            (!e.exit || e.exit.isAfter(monthEnd))
```
[the remark this answers is not in the document]
Comment (vonbearshark) [reply src/controllers/children/get.ts:82]: (The new addition is `&& (!e.exit || e.exit.isAfter(monthEnd))` is the diff is hard to parse

File: src/controllers/children/get.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jlhogan) [reply src/controllers/children/get.ts:85]: you're correct prettier gets run in pre-commit hook. you can set up whatever is your IDe to prettier on save tho

File: client/src/containers/Roster/Roster.tsx (hunk 1 above)
Comment (vonbearshark) [reply client/src/containers/Roster/Roster.tsx:98]: Yeah, it's pretty confusing, especially since we are filtering on both the front and back end. For the month filtering, we filter on the backend and return only the filtered children. But for the default/withdrawn breakdown, we do that here on the front end. It's the same endpoint, so the variable is actually like `allOrMonthFilteredChildren`.

Comment (vonbearshark) [reply client/src/containers/Roster/Roster.tsx:98]: > where the "active"(default) roster is just the roster with query month = this month

I would like that to be the case, but I am not sure it can be. Tried to capture the issue in https://github.com/ctoec/data-collection/pull/1231#discussion_r589897482 . If the `withdrawn` and `month` filters are mutually exclusive I don't see how to have `withdrawn` for given `month`. Unless they _aren't_ mutually exclusive, but then I think that might be a bigger refactor?

**Comment by vonbearshark** [comment]
Hey I am going to close this one out after convos IRL. Appreciate everyone's thoughts and 👀  but I think it warrants a larger scale refactor. Will try and capture the context in the Issue/Slack
